### PR TITLE
[engine] Speedy Machine should be invariant in its Question type

### PR DESCRIPTION
**Pull request by nickchapman-da**
Advances: #15882

Having the Speedy `Machine` covariant `[+Q]` in its question type plays very badly with the definition of its control field: `var control: Control[Q]` which is a contravariant instance. This is in fact only allowed by the scala type checker because it is declared `private[this]` (which seems somewhat of a nasty hack).

Better is to declare the speedy machine an invariant in `[Q]`, and fix the few places in the code which took advantage of the covariance. The _payoff_ is that we can get the _correct_ type for `setControl`.

This was: `final def setControl(x: Control[Nothing]): Unit`
But is now: `final def setControl(x: Control[Q]): Unit`

The improved type for `setControl` will be needed by the implementation of the new `Update` question: `NeedAuthority`

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -796,7 +800,7 @@ private[lf] object Speedy {
       track.reset()
     }
 
-    final def setControl(x: Control[Nothing]): Unit = {
+    final def setControl(x: Control[Q]): Unit = {
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:803]: This is the payoff :slightly_smiling_face:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -489,12 +489,11 @@ private[lf] object SBuiltin {
         args: util.ArrayList[SValue],
         machine: Machine[Q],
     ): Control.Value = {
-      val coid = getSContractId(args, 0).coid
-      machine match {
-        case _: PureMachine | _: ScenarioMachine =>
-          Control.Value(SOptional(Some(SText(coid))))
-        case _: UpdateMachine =>
-          Control.Value(SValue.SValue.None)
+      if (machine.isUpdateMachine) {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:492]: what is the problem with pattern machine ?

File: daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala [hunk 3]
```diff
@@ -347,17 +347,19 @@ object SBuiltinInterfaceTest {
       getContract: PartialFunction[Value.ContractId, Value.VersionedContractInstance],
       getKey: PartialFunction[GlobalKeyWithMaintainers, Value.ContractId],
   ): Try[Either[SError, SValue]] = {
-    val machine =
-      if (onLedger)
+    if (onLedger) {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala:350]: We should the usage of this function to be use only in "on-ledger". 
I suspect we used only once in "off-ledger" mode, where we can directly use  
`Speedy.Machine.fromPureSExpr(compiledBasePkgs, e).runPure()`

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ComparisonSBuiltinTest.scala [hunk 4]
```diff
@@ -687,7 +687,7 @@ class ComparisonSBuiltinTest extends AnyWordSpec with Matchers with TableDrivenP
         compiledPackages,
         SEApp(sexpr, (parties ++ contractIds).toArray),
       )
-    SpeedyTestLib.run(machine)
+    SpeedyTestLib.runPure(machine)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ComparisonSBuiltinTest.scala:690]:
```suggestion
    machine.runPure()
```

File: daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/daml/SBuiltinInterfaceTest.scala:350]: We should be able to kill `SpeedyTestLib.runPure`

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala [hunk 5]
```diff
@@ -36,6 +36,22 @@ private[speedy] object SpeedyTestLib {
 
   implicit def loggingContext: LoggingContext = LoggingContext.ForTesting
 
+  @throws[SError.SErrorCrash]
+  def runPure(machine: Speedy.Machine[Nothing]): Either[SError.SError, SValue] = {
+    runTxPure(machine) match {
+      case Left(e) => Left(e)
+      case Right(SResultFinal(v)) => Right(v)
+    }
+  }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala:45]:
```suggestion
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:492]: It requires that `Machine` is covariant in `Q`.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:492]: It is not really the pattern matching by itself that requires the covariance but more accurately the disjunctive pattern. 

```
 machine match {
        case _: UpdateMachine =>
           Control.Value(SValue.SValue.None)
         case _ => 
}
``` 

just work fine, it is clearer and make the method `isUpdateMachine` useless.

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:492]: Do you know the reason for this?
Or is it just  _Scala_ :shrug:

**Pull request merged**